File: daily-coding-problem/dcp-jvm/dcp-scala/src/main/scala/dcp/problems/_1_20/Problem5_FunctionalPair.scala
package dcp.problems._1_20

/**
 * This problem was asked by Jane Street.
 *
 * cons(a, b) constructs a pair, and car(pair) and cdr(pair) returns the first and last element of that pair.
 * For example, car(cons(3, 4)) returns 3, and cdr(cons(3, 4)) returns 4.
 *
 * Implement car and cdr
 */

object Problem5_FunctionalPair {

  type FunctionalPair = ((Int, Int) => Int) => Int

  def cons(a: Int, b: Int): FunctionalPair =
    (f: (Int, Int) => Int) => f(a, b)

  def car(pair: FunctionalPair): Int = {
    val f = (a: Int, _: Int) => a
    pair(f)
  }

  def cdr(pair: FunctionalPair): Int = {
    val f = (_: Int, b: Int) => b
    pair(f)
  }

}

File: daily-coding-problem/dcp-py/1_20/problem5_functional_pair.py
"""
This problem was asked by Jane Street.

cons(a, b) constructs a pair, and car(pair) and cdr(pair) returns the first and last element of that pair. For example,
car(cons(3, 4)) returns 3, and cdr(cons(3, 4)) returns 4.

```python
def cons(a, b):
    def pair(f):
        return f(a, b)
    return pair
```
Implement `car` and `cdr`.
"""


def cons(a, b):
    def pair(f):
        return f(a, b)

    return pair


def car(pair):
    def fun(x, _):
        return x

    return pair(fun)  # or return pair(lambda x, y: x)


def cdr(pair):
    def fun(_, y):
        return y

    return pair(fun)  # or return pair(lambda x, y: x)


assert car(cons(3, 4)) == 3
assert cdr(cons(3, 4)) == 4
